File: examples2d/collision_groups2.rs
use na::{Point2, Point3};
use rapier2d::dynamics::{JointSet, RigidBodyBuilder, RigidBodySet};
use rapier2d::geometry::{ColliderBuilder, ColliderSet, InteractionGroups};
use rapier_testbed2d::Testbed;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let joints = JointSet::new();

    /*
     * Ground
     */
    let ground_size = 5.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::new_static()
        .translation(0.0, -ground_height)
        .build();
    let floor_handle = bodies.insert(rigid_body);
    let collider = ColliderBuilder::cuboid(ground_size, ground_height).build();
    colliders.insert(collider, floor_handle, &mut bodies);

    /*
     * Setup groups
     */
    const GREEN_GROUP: InteractionGroups = InteractionGroups::new(0b01, 0b01);
    const BLUE_GROUP: InteractionGroups = InteractionGroups::new(0b10, 0b10);

    /*
     * A green floor that will collide with the GREEN group only.
     */
    let green_floor = ColliderBuilder::cuboid(1.0, 0.1)
        .translation(0.0, 1.0)
        .collision_groups(GREEN_GROUP)
        .build();
    let green_collider_handle = colliders.insert(green_floor, floor_handle, &mut bodies);

    testbed.set_collider_initial_color(green_collider_handle, Point3::new(0.0, 1.0, 0.0));

    /*
     * A blue floor that will collide with the BLUE group only.
     */
    let blue_floor = ColliderBuilder::cuboid(1.0, 0.1)
        .translation(0.0, 2.0)
        .collision_groups(BLUE_GROUP)
        .build();
    let blue_collider_handle = colliders.insert(blue_floor, floor_handle, &mut bodies);

    testbed.set_collider_initial_color(blue_collider_handle, Point3::new(0.0, 0.0, 1.0));

    /*
     * Create the cubes
     */
    let num = 8;
    let rad = 0.1;

    let shift = rad * 2.0;
    let centerx = shift * (num / 2) as f32;
    let centery = 2.5;

    for j in 0usize..4 {
        for i in 0..num {
            let x = i as f32 * shift - centerx;
            let y = j as f32 * shift + centery;

            // Alternate between the green and blue groups.
            let (group, color) = if i % 2 == 0 {
                (GREEN_GROUP, Point3::new(0.0, 1.0, 0.0))
            } else {
                (BLUE_GROUP, Point3::new(0.0, 0.0, 1.0))
            };

            let rigid_body = RigidBodyBuilder::new_dynamic().translation(x, y).build();
            let handle = bodies.insert(rigid_body);
            let collider = ColliderBuilder::cuboid(rad, rad)
                .collision_groups(group)
                .build();
            colliders.insert(collider, handle, &mut bodies);

            testbed.set_body_color(handle, color);
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, joints);
    testbed.look_at(Point2::new(0.0, 1.0), 100.0);
}

fn main() {
    let testbed = Testbed::from_builders(0, vec![("Boxes", init_world)]);
    testbed.run()
}

File: python/rapier-testbed/rapier_testbed/examples2/collision_groups2.py
"""Port of examples2d/collision_groups2.rs."""
from __future__ import annotations

import rapier2d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Collision groups"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 5.0
    ground_height = 0.1

    floor_handle = bodies.insert(rp.RigidBody.fixed().translation((0.0, -ground_height)))
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height), floor_handle, bodies
    )

    green_group = rp.InteractionGroups(
        memberships=rp.Group.GROUP_1, filter=rp.Group.GROUP_1
    )
    blue_group = rp.InteractionGroups(
        memberships=rp.Group.GROUP_2, filter=rp.Group.GROUP_2
    )

    green_floor = (
        rp.Collider.cuboid(1.0, 0.1)
        .translation((0.0, 1.0))
        .collision_groups(green_group)
    )
    colliders.insert_with_parent(green_floor, floor_handle, bodies)

    blue_floor = (
        rp.Collider.cuboid(1.0, 0.1)
        .translation((0.0, 2.0))
        .collision_groups(blue_group)
    )
    colliders.insert_with_parent(blue_floor, floor_handle, bodies)

    num = 8
    rad = 0.1
    shift = rad * 2.0
    centerx = shift * (num // 2)
    centery = 2.5

    for j in range(4):
        for i in range(num):
            x = i * shift - centerx
            y = j * shift + centery
            group = green_group if i % 2 == 0 else blue_group
            handle = bodies.insert(rp.RigidBody.dynamic().translation((x, y)))
            colliders.insert_with_parent(
                rp.Collider.cuboid(rad, rad).collision_groups(group), handle, bodies
            )

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 1.0), zoom=100.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
